Enabled per-project JVM mode configuration

hubertp opened the pull request:
### Pull Request Description

`package.yaml` now supports `jvm` field that takes a Boolean. If set to true, the project will be run in JVM mode even if Language Server's Native Image is available.
Also fixes a bug in runner when detecting JDK from JAVA_HOME on Windows.
Closes #12529.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

hubertp committed:
Enabled per-project JVM mode configuration

`package.yaml` now supports `jvm` field that takes a Boolean. If
set to true, the project will be run in JVM mode even if Language
Server's Native Image is available.

hubertp committed:
Merge branch 'develop' into wip/hubert/12529-config-jvm

hubertp committed:
unit test

Akirathan reviewed (commented):
Any chance we could do a simple test for this? Add a dummy project with `jvm = true` in the package descriptor, launch it via native engine, and look for "Started in JVM" in stdout? Something like that?

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -628,6 +628,7 @@ private void createNew(
             authors,
             nil(),
             "",
+            Option$.MODULE$.empty(),
```
This is the only change in `engine/runner`. Is that sufficient to switch to JVM mode when there is a project with `jvm: true` and one does:
```
enso --run the_project
```
? I am afraid that it is not sufficient.

JaroslavTulach commented on engine/language-server/src/test/scala/org/enso/languageserver/libraries/ComponentGroupsResolverSpec.scala:
```diff
@@ -280,7 +280,8 @@ object ComponentGroupsResolverSpec {
       preferLocalLibraries = true,
-      componentGroups      = Some(componentGroups)
+      componentGroups      = Some(componentGroups),
+      jvm                  = None
```
Is it important to have a tri-state value for `jvm`? Isn't binary `Boolean` enough?

hubertp replied:
It's not enough if we want to be backwards compatible with configs that don't have that field.

JaroslavTulach commented on lib/scala/pkg/src/main/scala/org/enso/pkg/Package.scala:
```diff
@@ -307,7 +308,8 @@ class PackageManager[F](implicit val fileSystem: FileSystem[F]) {
     license: String                          = "",
-    componentGroups: Option[ComponentGroups] = None
+    componentGroups: Option[ComponentGroups] = None,
+    jvm: Option[Boolean]                     = None
```
A tri-state value. Shouldn't we use just `Boolean` - e.g. binary state here?

hubertp replied:
I don't believe so, no.

JaroslavTulach commented on lib/scala/pkg/src/test/resources/Enso_Project_Archive_Test/package.yaml:
```diff
@@ -4,3 +4,4 @@ enso-version: default
 version: "0.0.1"
 author: "Enso Team <contact@enso.org>"
 maintainer: "Enso Team <contact@enso.org>"
+jvm: false
```
Why do we need `jvm: false` here? Isn't `false` the default when there is no `jvm:`?

JaroslavTulach commented on lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerController.scala:
```diff
@@ -79,15 +79,16 @@ class LanguageServerController(
+      engineVersion             = engineVersion,
+      jvmSettings               = distributionConfiguration.defaultJVMSettings,
+      jvmModeEnabled =
+        processConfig.jvmMode || project.jvmModeEnabled.getOrElse(false),
```
This is where the tri-state gets converted to binary state, right? Is `LanguageServerController` the right place to encode the default value? Shouldn't there be `project.isJvmMode(): Boolean`?

JaroslavTulach reviewed (changes requested):
- unless I am mistaken `jvm: true` has no effect on CLI execution
- we need CLI execution to behave the same as IDE execution (thus requesting changes)
- e.g. `enso` launcher needs to honor the `jvm:` value in `package.yaml`
- having a test to ensure that would be great
- other than that OK
- I just left few inline comments/opinions about tri-state and binary state

hubertp commented:
> * unless I am mistaken `jvm: true` has no effect on CLI execution

That is a valid point indeed.

hubertp committed:
CLI respects jvm option in package.yaml

Before launching Enso CLI, it will:
1) attempt to locate an appropriate `package.yaml` for the project
2) load config file and read `jvm` field
3) force JVM mode if config has the option set

This change also ensures that `Generic_JDBC_Tests` is now run in JVM
mode.

hubertp commented:
cli now also respects package.yaml configuration. `Generic_JDBC_Tests` now serves as a test - it would fail in NI but now it passes.

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -1334,6 +1336,52 @@ void stderr(String msg) {
+  private boolean isJvmModeEnabled(CommandLine line) {
+    var it = line.iterator();
+    String target = null;
+    while (target == null && it.hasNext()) {
```
Why loop manually and not use `line.hasOption` and `line.getOptionValue`?

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -1334,6 +1336,52 @@ void stderr(String msg) {
+    if (configFile == null) {
+      return false;
+    } else {
+      try (FileReader fileReader = new FileReader(configFile)) {
```
We are reading the config here and then the _running_ will read it again, right? Inefficiency (without functional impact).

hubertp replied:
Yes, I'm well aware. Not the only place where I think we have such small inefficiency, I believe.

JaroslavTulach commented on test/Generic_JDBC_Tests/package.yaml:
```diff
@@ -5,3 +5,4 @@ license: MIT
 author: enso-dev@enso.org
 maintainer: enso-dev@enso.org
 prefer-local-libraries: true
+jvm: true
```
> cli now also respects package.yaml configuration. `Generic_JDBC_Tests` now serves as a test - it would fail in NI but now it passes.

Should you modify [context.rs](https://github.com/enso-org/enso/blob/94d48dac7dcb4fdc1ca2d7c0b91b7941ad4e9231/build_tools/build/src/engine/context.rs#L718) (where all test projects executed by NI are listed) to also include `Generic_JDBC_Tests`?

hubertp replied:
Contrary to what one might expect, those aren't run anyway. See my [other PR](https://github.com/enso-org/enso/pull/12531) that fixes it.

hubertp committed:
Address PR feedback

hubertp committed:
typo

GregoryTravis reviewed (approved):
Approving library change

hubertp committed:
More tweaks to make CI happy

hubertp committed:
JAVA_HOME is set when running in JVM mode

Apparently we do build NI on some occassions, therefore requiring jdk
when running in JVM mode. But having JAVA_HOME set in the env is not
equivalent to it being available when running Generic_JDBC_Tests.

hubertp committed:
nit

hubertp committed:
More info to debug CI failures

hubertp committed:
Fix a bug in Windows OS detection

It doesn't look like this change ever worked on CI.
